x509-cert/src/name.rs

```diff
@@ -2,17 +2,181 @@
 
 use crate::attr::AttributeTypeAndValue;
 use alloc::vec::Vec;
+use const_oid::{
+    db::{rfc3280, rfc4519},
+    ObjectIdentifier,
+};
 use core::{fmt, str::FromStr};
-use der::{asn1::SetOfVec, Encode};
+use der::{
+    asn1::{Any, PrintableStringRef, SetOfVec},
+    Encode,
+};
 
 /// X.501 Name as defined in [RFC 5280 Section 4.1.2.4]. X.501 Name is used to represent distinguished names.
 ///
 /// ```text
 /// Name ::= CHOICE { rdnSequence  RDNSequence }
 /// ```
 ///
+/// # Example
+///
+/// ```
+/// use std::str::FromStr;
+/// use x509_cert::name::Name;
+///
+/// let subject = Name::from_str("CN=example.com").expect("correctly formatted subject");
+/// ```
+///
 /// [RFC 5280 Section 4.1.2.4]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.4
-pub type Name = RdnSequence;
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct Name(RdnSequence);
+
+impl_newtype!(Name, RdnSequence);
```

Comment:
Are the `From` impls defined here correct? That is, are all `RdnSequence`s valid `Name`s? Or are there rules that apply to `Name`s beyond what applies to `RdnSequence`?

Reply:
The `From` is necessary to comply with the escape hatch offered by the RFC:
```
When CAs have previously issued certificates with issuer fields with
attributes encoded using TeletexString, BMPString, or
UniversalString, then the CA MAY continue to use these encodings of
the DirectoryString to preserve backward compatibility.
```

Reply:
Is that the only restriction the RFC lifts? `From` in particular is a very generic trait and currently will let the user construct literally whatever `Name` they want, due to `RdnSequence` being completely transparent.

So I ask again: what restrictions are placed on a `Name` by the RFC? And do any of them *not* have backwards-compatibility escape hatches in the RFC?

Reply:
I removed the`From<RdnSequence>` and placed an helper under hazmat.